pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java

```diff
@@ -28,9 +28,218 @@
+      if (numChildFilterOperators == 0) {
+        // Return empty filter operator if all child filter operators's result is empty
+        return EmptyFilterOperator.getInstance();
+      } else if (numChildFilterOperators == 1) {
+        // Return the child filter operator if only one left
+        return childFilterOperators.get(0);
+      } else {
+        // Return the OR filter operator with child filter operators
+        return new OrFilterOperator(childFilterOperators, numDocs);
+      }
+    }
+
+    @Override
+    public BaseFilterOperator getNotFilterOperator(QueryContext queryContext, BaseFilterOperator filterOperator,
+        int numDocs) {
+      if (filterOperator.isResultMatchingAll()) {
+        return EmptyFilterOperator.getInstance();
+      } else if (filterOperator.isResultEmpty()) {
+        return new MatchAllFilterOperator(numDocs);
+      }
+
+      return new NotFilterOperator(filterOperator, numDocs);
+    }
+
+    @Override
+    public void reorderAndFilterChildOperators(QueryContext queryContext, List<BaseFilterOperator> filterOperators) {
+      filterOperators.sort(new Comparator<BaseFilterOperator>() {
+        @Override
+        public int compare(BaseFilterOperator o1, BaseFilterOperator o2) {
+          return getPriority(o1) - getPriority(o2);
+        }
+
+        int getPriority(BaseFilterOperator filterOperator) {
+          if (filterOperator instanceof SortedIndexBasedFilterOperator) {
+            return 0;
+          }
+          if (filterOperator instanceof BitmapBasedFilterOperator) {
+            return 1;
+          }
+          if (filterOperator instanceof RangeIndexBasedFilterOperator
+              || filterOperator instanceof TextContainsFilterOperator
+              || filterOperator instanceof TextMatchFilterOperator || filterOperator instanceof JsonMatchFilterOperator
+              || filterOperator instanceof H3IndexFilterOperator
+              || filterOperator instanceof H3InclusionIndexFilterOperator) {
+            return 2;
+          }
+          if (filterOperator instanceof AndFilterOperator) {
+            return 3;
+          }
+          if (filterOperator instanceof OrFilterOperator) {
+            return 4;
+          }
+          if (filterOperator instanceof NotFilterOperator) {
+            return getPriority(((NotFilterOperator) filterOperator).getChildFilterOperator());
+          }
+          if (filterOperator instanceof ScanBasedFilterOperator) {
+            return getScanBasedFilterPriority(queryContext, (ScanBasedFilterOperator) filterOperator, 5);
+          }
+          if (filterOperator instanceof ExpressionFilterOperator) {
+            return 10;
+          }
+          throw new IllegalStateException(filterOperator.getClass().getSimpleName()
+              + " should not be reordered, remove it from the list before calling this method");
+        }
+      });
+    }
+
+    public static int getScanBasedFilterPriority(QueryContext queryContext,
```

Comment:
I don't get why we cannot keep this method as `protected`. IMO both `reorderAndFilterChildOperators()` and `getScanBasedFilterPriority()` should be protected

Reply:
Yes, we can do it given that it is a class.